controller/volume_controller.go

```diff
@@ -771,33 +790,101 @@ func (vc *VolumeController) cleanupFailedToScheduledReplicas(v *longhorn.Volume,
 
 func (vc *VolumeController) cleanupExtraHealthyReplicas(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica) (err error) {
 	healthyCount := vc.getHealthyReplicaCount(rs)
+	if healthyCount <= v.Spec.NumberOfReplicas {
+		return nil
+	}
 
-	if healthyCount > v.Spec.NumberOfReplicas {
-		for _, r := range rs {
-			// Clean up eviction requested replica
-			if r.Status.EvictionRequested {
-				if err := vc.deleteReplica(r, rs); err != nil {
-					vc.eventRecorder.Eventf(v, v1.EventTypeWarning,
-						EventReasonFailedEviction,
-						"volume %v failed to evict replica %v",
-						v.Name, r.Name)
-					return err
-				}
-				logrus.Debugf("Evicted replica %v", r.Name)
-				return nil
+	var cleaned bool
+	cleaned, err = vc.cleanupEvictionRequestedReplicas(v, rs)
+	if err != nil {
+		return err
+	}
+	if cleaned {
+		return nil
+	}
+
+	cleaned, err = vc.cleanupAutoBalancedReplicas(v, rs)
+	if err != nil {
+		return err
+	}
+	if cleaned {
+		return nil
+	}
+
+	cleaned, err = vc.cleanupDataLocalityReplicas(v, e, rs)
+	if err != nil {
+		return err
+	}
+	if cleaned {
+		return nil
+	}
+
+	return nil
```

Comment:
nit: err will be nil when cleaned == true, no need for the spaghetti
```suggestion
	if cleaned, err = vc.cleanupEvictionRequestedReplicas(v, rs); err != nil || cleaned {
		return err
	}
	
	if cleaned, err = vc.cleanupAutoBalancedReplicas(v, rs); err != nil || cleaned {
		return err
	}
	
	if cleaned, err = vc.cleanupDataLocalityReplicas(v, e, rs); err != nil || cleaned {
		return err
	}
	
	return nil
```

Reply:
Thanks! fixed.